### ISSUE TITLE
[building-rollup] babel-helpers moved to a separate file?

### ISSUE THREAD

**Issue by rwestlund** [open-wc/open-wc#1625]
It looks like building-rollup is now putting the babel-helpers in a separate file, as an asset. It seems like a mistake to have a second tiny output file when my whole app's JS was bundled in one output file before.

I removed `bundledBabelHelpers()` from the rollup output plugins and `babelPluginBundledHelpers` from the babel plugins in `createBabelConfigRollupGenerate()`. It appears to be doing what I want; I'm back to one output bundle, with the helpers defined once at the top.

Am I missing something? Is there a reason for all this code that's separating the helpers?

**Comment by LarsDenBakker** [comment]
This is somewhat intentional. We are using a feature from rollup which runs babel on the generated output chunks rather than transformations on individual files. This also allows creating multiple outputs from a single build. We use this for older browsers too.

By default babel puts helpers on top of a file. When you run rollup with babel per file, there is an option to deduplicate these helpers. When running babel on the generated chunks, this option isn't there. 

Duplicating the helpers for each chunk is bad, so I created a solution for now which creates a separate chunk with the helpers. Ideally this chunk is merged into existing chunks, but that would be quite hard to do since rollup has done all of it's logic already.

**Comment by rwestlund** [comment]
I must be misunderstanding something, because I think I am both running babel on the output and also getting deduplicated helpers in the main bundle.

I built this config after reading the building-rollup source. My goal is to do the same build, but with the helpers in the main bundle, and it appears to be working. The two relevant changes are marked below:

```js                      
import html from '@open-wc/rollup-plugin-html';                                 
import path from 'path';                                                        
import resolve from '@rollup/plugin-node-resolve';                                                 
import { terser } from 'rollup-plugin-terser';                                  
import polyfillsLoader from '@open-wc/rollup-plugin-polyfills-loader';          
import babel from 'rollup-plugin-babel';                                        
import { babelConfigRollupGenerate, createBabelConfigRollupBuild } from '@open-wc/building-rollup/src/babel/babel-configs';
import { bundledBabelHelpers } from '@open-wc/building-rollup/src/babel/rollup-plugin-bundled-babel-helpers';

// CHANGE #1: Remove babelPluginBundledHelpers.                                            
babelConfigRollupGenerate.plugins = babelConfigRollupGenerate.plugins.slice(1); 
                                                                                
const c = {                                                                     
    input: path.resolve(__dirname, "index.html"),                               
    preserveEntrySignatures: false,                                             
    treeshake: true,                                                            
    output: {                                                                   
        format: "es",                                                           
        dir: "dist",                                                            
        entryFileNames: "[name]-[hash].js",                                     
        plugins: [                                                         
            babel.generated(babelConfigRollupGenerate), 
            // CHANGE #2: Remove this.
            //bundledBabelHelpers,                                                                                       
        ],                                                                      
    },                                                                          
    plugins: [                                                                  
        resolve(),                                                              
        babel(createBabelConfigRollupBuild()),                                  
        terser({ output: { comments: false } }),                              
        html({                                                                  
            minify: true,                                                       
            inject: false,                                                      
        }),                                                                     
        polyfillsLoader({                                                       
            polyfills: {                                                        
                fetch: true,                                                    
                abortController: true,                                          
                coreJs: true,                                                   
                webcomponents: true,                                            
            },                                                                  
        }),                                                                                                                                
    ],                                                                                                               
};                                         
export default c;
```
With the default build, I see `objectSpread2` defined in a separate file (the only helper that gets pulled in). With this config, I see `objectSpread` defined only once at the top of the main bundle.

**Comment by LarsDenBakker** [comment]
If you have multiple chunks using object spread, it will show up in each chunk.

For some projects this will be a problem, for others it wont. The goal is to try and find a solution that solves this for all projects.

**Comment by rwestlund** [comment]
Oh, I thought the concern was helpers being inlined in each class. Once per chunk/bundle isn't bad. It might be nice to have a config option to toggle this behavior, as packing SPAs into one bundle seems common.

Thanks for the explanation!

**Comment by LarsDenBakker** [comment]
If you do any form of lazy loading you will end up with more than onr chunk, where this is very bad. Especially for the legacy build.

**Comment by stephenwade** [comment]
> Duplicating the helpers for each chunk is bad

Why is it bad? I would think the extra wait time from loading extra files is worse than the extra bytes from duplicated helpers.

**Comment by LarsDenBakker** [comment]
Duplicate babel helpers results in many kbs, and projects which do a lot of lazy loading will have many chunks. So the duplication is very bad there. But it depends on the type of project of course.

Loading one or two files in parallel does not make any difference, HTTP2 can handle that fine up to a certain amount of parallel requests. In fact the babel helpers will be able to benefit from long term caching. 

Still, there is room for improvement here especially for projects that don't do any dynamic imports. This is high on the priority list to solve.

**Comment by stephenwade** [comment]
Ah, that makes sense.

One way to improve would be to add a preload tag in the HTML header for the babel chunks. Currently the main chunk is in a preload tag, but the browser has to download and parse that JS file before it learns about the other JS file, so even with HTTP2 there is that delay.

Would that be easy to do?

**Comment by LarsDenBakker** [comment]
Sorry this took much longer to get around to than I wanted. 

We're not avoiding the helper chunk entirely for the modern build.
